packages/graph/hash_graph/lib/graph/src/store/query/ontology.rs

```diff
@@ -0,0 +1,74 @@
+use std::{fmt, fmt::Formatter, marker::PhantomData};
+
+use crate::ontology::types::{uri::BaseUri, DataType, EntityType, LinkType, PropertyType};
+
+pub type DataTypeQuery<'q> = OntologyQuery<'q, DataType>;
+pub type PropertyTypeQuery<'q> = OntologyQuery<'q, PropertyType>;
+pub type LinkTypeQuery<'q> = OntologyQuery<'q, LinkType>;
+pub type EntityTypeQuery<'q> = OntologyQuery<'q, EntityType>;
+
+#[derive(Debug, Copy, Clone)]
+pub enum OntologyVersion {
+    Exact(u32),
+    Latest,
+}
+
+pub struct OntologyQuery<'q, T> {
+    _marker: PhantomData<fn() -> T>,
+    uri: Option<&'q BaseUri>,
+    version: Option<OntologyVersion>,
+}
+
+impl<T> fmt::Debug for OntologyQuery<'_, T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("OntologyQuery")
+            .field("uri", &self.uri())
+            .field("version", &self.version())
+            .finish()
+    }
+}
+
+impl<T> OntologyQuery<'_, T> {
+    #[must_use]
+    pub const fn new() -> Self {
+        Self {
+            _marker: PhantomData,
+            uri: None,
+            version: None,
+        }
+    }
+}
+
+/// Methods for building up a query.
+impl<'q, T> OntologyQuery<'q, T> {
+    #[must_use]
+    pub const fn with_uri(mut self, uri: &'q BaseUri) -> Self {
+        self.uri = Some(uri);
+        self
+    }
+
+    #[must_use]
+    pub const fn with_version(mut self, version: u32) -> Self {
+        self.version = Some(OntologyVersion::Exact(version));
+        self
+    }
+
+    #[must_use]
+    pub const fn with_latest_version(mut self) -> Self {
+        self.version = Some(OntologyVersion::Latest);
+        self
+    }
+}
```

Comment:
Instead of `with_*` we probably want to use `by_*`. This enables us to add `with_*` methods later to change the return type.
We need to create the query (`by_x().by_y().by_z()`) without changing any generic parameters, as this is required for deserialization of a query. However, `with_*` may introduce additional parameters, which then can be used to determine the return type (e.g. `DataTypeQuery::new().by_uri(...).by_latest_version().with_account_id()` will return the account id of the specified URI+version)

Reply:
I renamed `with_*` to `by_*` in 202bd3266fcd5923071ccbaef3e9bd0088307e83.